internal/collector/processor/queued_processor.go

```diff
@@ -0,0 +1,146 @@
+	/* TODO: (@pauloja) not doing metrics for now
+	metrics                  *cApp.SpanProcessorMetrics
+	batchMetrics             *processorBatchMetrics
+	*/
+	logger                   *zap.Logger
+	sender                   SpanProcessor
+	numWorkers               int
+	retryOnProcessingFailure bool
+	backoffDelay             time.Duration
+	stopCh                   chan struct{}
+}
+
+var _ SpanProcessor = (*queuedSpanProcessor)(nil)
+
+type queueItem struct {
+	queuedTime time.Time
+	batch      *agenttracepb.ExportTraceServiceRequest
+	spanFormat string
+}
+
+// NewQueuedSpanProcessor returns a span processor that maintains a bounded
+// in-memory queue of span batches, and sends out span batches using the
+// provided sender
+func NewQueuedSpanProcessor(sender SpanProcessor, opts ...Option) SpanProcessor {
+	sp := newQueuedSpanProcessor(sender, opts...)
+
+	sp.queue.StartConsumers(sp.numWorkers, func(item interface{}) {
+		value := item.(*queueItem)
+		sp.processItemFromQueue(value)
+	})
+
+	return sp
+}
+
+func newQueuedSpanProcessor(
+	sender SpanProcessor,
+	opts ...Option,
+) *queuedSpanProcessor {
+	options := Options.apply(opts...)
+	boundedQueue := queue.NewBoundedQueue(options.queueSize, func(item interface{}) {})
+	return &queuedSpanProcessor{
+		queue:                    boundedQueue,
+		logger:                   options.logger,
+		numWorkers:               options.numWorkers,
+		sender:                   sender,
+		retryOnProcessingFailure: options.retryOnProcessingFailure,
+		backoffDelay:             options.backoffDelay,
+		stopCh:                   make(chan struct{}),
+	}
+}
+
+// Stop halts the span processor and all its go-routines.
+func (sp *queuedSpanProcessor) Stop() {
+	close(sp.stopCh)
+	sp.queue.Stop()
+}
+
+// ProcessSpans implements the SpanProcessor interface
+func (sp *queuedSpanProcessor) ProcessSpans(batch *agenttracepb.ExportTraceServiceRequest, spanFormat string) (uint64, error) {
+	ok := sp.enqueueSpanBatch(batch, spanFormat)
+	if !ok {
+		return uint64(len(batch.Spans)), nil
+	}
+	return 0, nil
+}
+
+func (sp *queuedSpanProcessor) enqueueSpanBatch(batch *agenttracepb.ExportTraceServiceRequest, spanFormat string) bool {
+	item := &queueItem{
+		queuedTime: time.Now(),
+		batch:      batch,
+		spanFormat: spanFormat,
+	}
+	addedToQueue := sp.queue.Produce(item)
+	if !addedToQueue {
+		sp.onItemDropped(item)
+	}
+	return addedToQueue
+}
+
+func (sp *queuedSpanProcessor) processItemFromQueue(item *queueItem) {
+	// TODO: @(pjanotti) metrics: startTime := time.Now()
+	// TODO:
+	_, err := sp.sender.ProcessSpans(item.batch, item.spanFormat)
+	if err != nil {
+		batchSize := len(item.batch.Spans)
+		if !sp.retryOnProcessingFailure {
+			// throw away the batch
+			sp.logger.Error("Failed to process batch, discarding", zap.Int("batch-size", batchSize))
+			sp.onItemDropped(item)
+		} else {
+			// try to put it back at the end of queue for retry at a later time
+			addedToQueue := sp.queue.Produce(item)
```

Comment:
I think we should put a TODO - rather than put this at the back of the queue, next time we should retry with the same batch until it is sent successfully or processor is shut down (on shut down we should return it to back of queue).
this way we maintain the ordering in the queue (except for this one batch)

Reply:
Agreed, we need to be careful if the source of the failure is somehow related to the batch itself, in other words: is it a recoverable error or not.